### PR TITLE
fix: upsert replicas in cases where they could be deleted

### LINKED ISSUE #8858: replicasync: deleting temporarily unhealthy replica breaks node even after it recovers
with multiple Coderd replicas, each Coderd periodically runs a query that deletes replicas that haven't updated in 3x the configured interval (3 x 5s = 15s by default).  If a replica becomes temporarily unhealthy or disconnected from the DB, other Coderds will delete its row.  Then, if it reconnects, it attempts a SQL `UPDATE` on its row, which fails because the row is deleted.  It fails like this in a loop forever.

I have personally observed this in a test environment.

We should modify this `UPDATE` query to be an "upsert", e.g. `INSERT INTO ... ON CONFLICT ... DO UPDATE`

A workaround is to restart the Coderd so that it re-inserts itself into the table.

Finally, the log describing this problem is a "WARNING" but it should definitely be an "ERROR" --- if we can't update the replicas table it's a big problem.

### PR DESCRIPTION
**Pull request by f0ssel**
While I'm looking for things to work on I'm just tackling bugs

Closes https://github.com/coder/coder/issues/8858

### CODE REVIEW

**Comment by f0ssel** [comment]
I have read the CLA Document and I hereby sign the CLA

File: coderd/database/queries/replicas.sql [hunk 1]
```diff
@@ -18,19 +18,30 @@ INSERT INTO replicas (
...
 
--- name: UpdateReplica :one
-UPDATE replicas SET
+-- name: UpsertReplica :one
+INSERT INTO replicas (
+    id,
+    updated_at,
+    created_at,
+    started_at,
+    hostname,
```
Comment (spikecurtis) [inline coderd/database/queries/replicas.sql:27]: I think we should include `stopped_at` on the `INSERT` as well.  We'd expect it not to be set when a replica inserts itself the first time, but having an asymmetry between `INSERT` and `UPDATE` on an "upsert" can lead to confusing results.

For example, if one Coderd deletes a row, then another upserts it with `stopped_at` set to `now()`, we'd reinsert it without the `stopped_at`.

File: coderd/database/dbmem/dbmem.go [hunk 2]
```diff
@@ -5856,7 +5856,7 @@ func (q *FakeQuerier) UpdateProvisionerJobWithCompleteByID(_ context.Context, ar
 	return sql.ErrNoRows
 }
 
-func (q *FakeQuerier) UpdateReplica(_ context.Context, arg database.UpdateReplicaParams) (database.Replica, error) {
+func (q *FakeQuerier) UpsertReplica(_ context.Context, arg database.UpsertReplicaParams) (database.Replica, error) {
```
Comment (spikecurtis) [inline coderd/database/dbmem/dbmem.go:5859]: This implementation needs to be changed to insert if not present!

File: coderd/database/models.go [hunk 3]
```diff
@@ -1,6 +1,6 @@
 // Code generated by sqlc. DO NOT EDIT.
 // versions:
-//   sqlc v1.20.0
+//   sqlc v1.24.0
```
Comment (spikecurtis) [inline coderd/database/models.go:3]: You'll need to get your environment matching CI --- easiest is to use the Coder on Coder template, or Nix.

**Review by spikecurtis — CHANGES_REQUESTED** [review]
I'd also like to see a unit test that reproduces #8858 and shows that this fixes it.

**Pull request closed without merging**